### ISSUE TITLE
Possible bug in attention seq2seq

### ISSUE THREAD

**Issue by wchan** [tensorflow/tensorflow#512]
source file is seq2seq.py

    def attention(query):
      """Put attention masks on hidden using hidden_features and query."""
      ds = []  # Results of attention reads will be stored here.
      for a in xrange(num_heads):
        with vs.variable_scope("Attention_%d" % a): 
          y = rnn_cell.linear(query, attention_vec_size, True)
          y = array_ops.reshape(y, [-1, 1, 1, attention_vec_size])
          # Attention mask is a softmax of v^T * tanh(...).
          s = math_ops.reduce_sum(
              v[a] * math_ops.tanh(hidden_features[a] + y), [2, 3]) 
          a = nn_ops.softmax(s)
          # Now calculate the attention-weighted vector d.
          d = math_ops.reduce_sum(
              array_ops.reshape(a, [-1, attn_length, 1, 1]) * hidden,
              [1, 2]) 
          ds.append(array_ops.reshape(d, [-1, attn_size]))
      return ds
                                       

at line 440, you compute the energies, ok… then at line 441 you normalize the energies… you need to do a -FLT_MAX mask on only the part of the encoder sequence (i.e., you don’t want to pay attention to “empty” part of the sequence)... I don't see the masking op? Or am i missing something?

I can wrap an op patch for this if nobody is working on this (let me know).

Thanks all!

**Comment by lukaszkaiser** [comment]
I was thinking about it and there were 2 things that stopped me from working on it. The first and most important one is that I get good results without any special handling for masks, just as it is. Are you sure that a -FLT_MAX really helps, did you try without it? (I think the net just learns to never attend to the padding easily.) The second reason was that one day, when "while" works ok, we might not need to pad at all -- and I did not want to add complexity to the API that is not applicable to that case.

That said, if you find it important and want to do it -- by all means, send a patch :).

**Comment by LeavesBreathe** [comment]
> I get good results without any special handling for masks, just as it is

I too have been using this for about a month, and I also get good results just as it is. I think it learns to never attend to padding as that is a simple if statement. In keras, we also talked about this, and concluded that masking was not necessary because the network learns to ignore padding.

**Comment by ogrisel** [comment]
But for sequence with a lot of padding, this will result in an increase in the softmax attention temperature. I think this equivalent to having an additional implicit hyperparameter that changes for different input sequence lengths. The network might be somewhat robust to it as long as the input sequences do not vary too much but it is probably misleading for the user to have something that does not follow the mathematical description of most papers in the literature. That might be the source of hard to debug confusion when trying to reproduce the results of experiments done by other researchers with other tools.

**Comment by lukaszkaiser** [comment]
Indeed, it's a good argument. Still, as said before, the right solution to be close to the formalism is probably to use "while" (i.e., make the graph dynamically unroll) and then avoid padding altogether. We're working on this and did not want to polute the API with elements that will not be used then (as the padding symbol). But it will take a while before it's ready...

**Comment by wchan** [comment]
@lukaszkaiser even w/ While, wouldnt this still be an issue? esp if the sequences in the minibatch have high variances in sequence length.

**Comment by LeavesBreathe** [comment]
>The network might be somewhat robust to it as long as the input sequences do not vary too much  @ogrisel , if you apply enough different types of bucket sizes then you wouldn't really run into this problem. All the sequences would roughly be within 5 or 7 timesteps (assuming you're doing NLP seq2seq).

**Comment by lukaszkaiser** [comment]
You're right Will - if there is variance in minibatches this might always be a problem. And let's not forget there is another problem with padding - we start the encoder from the first batch element, so the state goes through a bunch of pad symbols before the real sentence starts. Both could be corrected if the padding symbol was passed to the function and it constructed appropriate masks - we should probably do that. Are you guys willing to take a shot at it? (I'll probably not manage to do it before leaving for Christmas.)

**Comment by wchan** [comment]
fyi, i wrote a simple op for this ... needs to be refactored and polished up before a patch can be submitted.

impl can be found here for those who need it before its pushed upstream:
https://github.com/wchan/tensorflow/blob/master/tensorflow/core/kernels/attention_mask_op_test.cc

**Comment by wchan** [comment]
FYI, I've added "MedianMask" and "WindowMask" (same path as above github path); see: http://arxiv.org/pdf/1508.04395.pdf

however, I still haven't been able to reduce their results.

**Comment by girving** [comment]
@lukaszkaiser: Can this issue be closed?

**Comment by lukaszkaiser** [comment]
Yes - let's close this and leave the seq2seq attention model as is (many people are using it successfully). We might implement another attention model with more features at some point, but that's a different issue.

**Comment by SivanKe** [comment]
I have to say that fixing this bug helped my results a lot!
I train a model of cnn + attention for OCR, and after adding a change to fix this bug my word error rate improved noticeably.
I know that you are implementing a new dynamic attention mechanism, but until it is stable (documented and tested) many people will probably stay with the old mechanism, so it can be nice if you could write a documentation of the bug in your API.
